aws-android-sdk-iot/src/main/java/com/amazonaws/mobileconnectors/iot/AWSIotSslUtility.java

```diff
@@ -19,26 +19,54 @@
 import java.security.KeyStore;
 import java.security.KeyStoreException;
 import java.security.NoSuchAlgorithmException;
+import java.security.NoSuchProviderException;
 import java.security.SecureRandom;
+import java.security.Security;
 import java.security.UnrecoverableKeyException;
 
 import javax.net.ssl.KeyManager;
 import javax.net.ssl.KeyManagerFactory;
 import javax.net.ssl.SSLContext;
 import javax.net.ssl.SSLSocketFactory;
+import org.conscrypt.OpenSSLProvider;
 
 /**
  * AWSIoTSslUtility Utility class for creating an SSL Socket Factory from
  * certs/keys in keystore.
  */
 final class AWSIotSslUtility {
 
+    protected static final String[] ALPN_EXTENSION = {"x-amzn-mqtt-ca"};
+    private static final String TLS_V_1_2 = "TLSv1.2";
+    protected static int portNumber = 8883;
+
     /**
      * Utility class.
      */
     private AWSIotSslUtility() {
     }
 
+    /**
+     * Creates a socket factory given a keystore and a portNumber.
+     *
+     * @param keyStore keystore containing a certificate and private key for
+     *            used in creating a secured socket.
+     * @param portNum Port number used for connecting to Iot
+     * @return a socket factory for use in creating a secured socket.
+     * @throws NoSuchAlgorithmException when TLS 1.2 is not available.
+     * @throws UnrecoverableKeyException when the private key cannot be
+     *             recovered. Ususally a bad keystore password.
+     * @throws KeyStoreException when keystore cannot be created.
+     * @throws KeyManagementException when SSL context cannot be created by key
+     *             manager.
+     */
+    protected static SSLSocketFactory getSocketFactoryWithKeyStore(KeyStore keyStore, int portNum)
+            throws NoSuchAlgorithmException, UnrecoverableKeyException, KeyStoreException,
+            KeyManagementException {
+        portNumber = portNum;
```

Comment:
You may need a syncrhonization mechanism to protect the assignment of portNumber. For example, when the app has multiple threads T1 and T2; T1 is calling connect and T2 is calling connectUsingALPN. There will be a race condition when the assignment of portNumber is not protected by a lock since portNumber is global state.